### PR TITLE
Allow loading custom .dll file(s)

### PR DESCRIPTION
**Pull request by wipe2238**
Filenames are taken from config, and loaded after sfall initialization. `[Debugging]` isn't exactly a perfect place for it, but i couldn't find anything better.

### CODE REVIEW

**Comment by NovaRain** [comment]
Is there an example dll for testing?

**Comment by wipe2238** [comment]
It was tested with [fo1in2.dll](https://github.com/rotators/Fo1in2/blob/master/Fallout2/Fallout1in2/fo1in2.dll) right before your commit with red dots, worked fine. There's too much asm involved now for me to say if both things won't conflict now :(

Unless you mean you'd want something to put into modders pack for people to do quick tests?

**Comment by NovaRain** [comment]
I was asking for myself because obviously I need to test the option before merging PR.

**Comment by FakelsHub** [comment]
> Unless you mean you'd want something to put into modders pack for people to do quick tests?

Quick tests are done directly in sfall.
there is already a lot of things made, and don't have to re-write the implementation (on ASM) to access the engine functions.

I assume this will come in handy for some external plug-in mods/modules, like Plug&Play :)

**Comment by wipe2238** [comment]
https://rotators.fodev.net/wipe/sfall/sfall-273.dll
`MessageBox()` on `DLL_PROCESS_ATTACH`/`DLL_PROCESS_DETACH`

File: sfall/Modules/LoadDll.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include <vector>
+
+#include "..\main.h"
+#include "LoadDll.h"
+
+namespace sfall
+{
+
+void LoadDll::init() {
+	std::vector<std::string> names = GetIniList("Debugging", "LoadDll", "", 512, ',', ddrawIni);
```
Comment (FakelsHub) [inline sfall/Modules/LoadDll.cpp]: this should be moved to the `[Main]` section

Comment (wipe2238) [reply sfall/Modules/LoadDll.cpp]: Done.

**Comment by FakelsHub** [comment]
> configuration moved to [Main]

option in ddraw )

**Comment by FakelsHub** [comment]
t is necessary to add validation methods so that it would not be possible to load any third-party library that is not intended for fallout.

**Comment by wipe2238** [comment]
> validation methods

- `LoadLibrary()`
- `GetProcAddress( "SFALL_EXTENSION")` or any other fancy name
  - if missing, `FreeLibrary()` and go to another dll

Should be enough without too much wizardy involved.

**Comment by FakelsHub** [comment]
Why was this closed and deleted?
You could make a separate library for F1inF2, to connect to sfall, rather than doing **Voodoo magic** in your scripts.

**Comment by wipe2238** [comment]
> Why was this closed and deleted?

>  NovaRain requested a review from phobos2077 on 2 Jan

I'd honestly be happier with simple _yes_ or _no_ said right away instead of PR hanging for gods know how long. There's lot of work needed to make dll "plugins" actually useful, and working on something what might be rejected half year from now... meh.

> You could make a separate library for F1inF2, to connect to sfall, rather than doing Voodoo magic in your scripts.

At this point i think improving `call_offset` / `write_int` etc. is better solution than custom dll (both _ddraw_ and "plugin"), not just for 1@2, but in general. With just few restrictions relaxed and small improvements here and there, mods could add whatever crazy stuff they need without need for any kind of custom dll - _ddraw_ or "plugin" - if such changes are added to vanilla sfall, that is. We (slowly) started working on that, but it's not tested/mature enough.

**Comment by FakelsHub** [comment]
> At this point i think improving call_offset / write_int etc. is better solution than custom dll 

Writing something to the memory of the f2 process, outside of sfall, is a bad decision.
Calling function or reading yet normal.

Any mod written in this way can fail, since such hacks are not controlled, and at any time sfall can replace something in the engine code.

**Comment by wipe2238** [comment]
> Writing something to the memory of the f2 process, outside of sfall, is a bad decision.

While i partially agree, having them in scripts allows to easily turn them on/off without need to update sfall every time. If it's broken, just delete macro, recompile script, and start slapping whoever wrote it. 

> Any mod written in this way can fail, since such hacks are not controlled, and at any time sfall can replace something in the engine code.

True. I saw `CheckConflict()` somewhere which could (?) help detecting such cases if it would track `write_*` from scripts as well. It's available only for `Dev` configuration now, but it could be `[Debugging]` option in regular build, for people playing with fire.

In some cases script-hacks are one of last chances to add something. Even if mod and sfall teams disagree what is a feature and what is a bug (#255), mod still have unsafe scripting as a tool to implement features it needs.

So can we focus on making sure sfall and script hacks are safe(-ish) to use together, and improve related functions a bit? That would help with making sure our mod runs on vanilla sfall, which is pretty important point on our wishlist. We can sort out later which parts of voodoo.h can be moved into sfall and how it should be done.

**Pull request closed without merging**